[elastic-agent] Re-introduce instrumentation

stuartnelson3 opened the pull request:
cf. https://github.com/elastic/beats/pull/29031 for original description.

seems the issue might have been forgetting to add credentials when starting up the gRPC servers: https://github.com/elastic/beats/commit/76383672ae27867da4c6445e69d9eac604cae335

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > There is a new build on-going so the previous on-going builds have been aborted.
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30385/detail/PR-30385/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30385/detail/PR-30385/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30385/detail/PR-30385/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30385/detail/PR-30385/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-02-14T18:06:55.225Z&rangeTo=2022-02-14T18:26:55.225Z&transactionName=Beats/beats/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=ad6815f1c4ef3e33c1d629747842258a&transactionId=52d1caf9a384f029)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-02-14T18:16:55.225+0000


* Duration: 24 min 13 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 3  |
| Skipped      | 0 |
| Total        | 3   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->











### Steps errors [![2](https://img.shields.io/badge/2%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30385/detail/PR-30385/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `metricbeat-lint - make -C metricbeat check;make -C metricbeat update;make -C x-pack/metricbeat che`
<ul>
<li>Took 3 min 53 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-30385/runs/1/steps/5532/log/?start=0">here</a></li>
<li>Description: <code>make -C metricbeat check;make -C metricbeat update;make -C x-pack/metricbeat check;make -C x-pack/metricbeat update;make check-no-changes;</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-30385/runs/1/steps/4456/log/?start=0">here</a></li>
<li>Description: <code>Error 'org.jenkinsci.plugins.workflow.steps.FlowInterruptedException'</code></l1>
</ul>
  
  </p>
  </details>


## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
Pinging @elastic/elastic-agent-control-plane  (Team:Elastic-Agent-Control-Plane)

stuartnelson3 commented:
/test

AndersonQ commented on x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go:
```diff
@@ -65,38 +68,51 @@ func NewController(
+func (e *Controller) Update(ctx context.Context, c *config.Config) (rErr error) {
+	span, ctx := apm.StartSpan(ctx, "update", "app.internal")
+	defer func() {
+		if rErr != nil {
+			apm.CaptureError(ctx, rErr).Send()
+		}
```
[Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, rErr).Send()
```

AndersonQ commented on x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go:
```diff
@@ -74,7 +76,15 @@ func (ad *ActionDispatcher) key(a fleetapi.Action) string {
+func (ad *ActionDispatcher) Dispatch(ctx context.Context, acker store.FleetAcker, actions ...fleetapi.Action) (err error) {
+	span, ctx := apm.StartSpan(ctx, "dispatch", "app.internal")
+	defer func() {
+		if err != nil {
+			apm.CaptureError(ctx, err).Send()
+		}
```
[Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

AndersonQ commented on x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go:
```diff
@@ -105,25 +121,41 @@ func (e *Controller) Update(c *config.Config) error {
+	var err error
+	span, ctx := apm.StartSpan(ctx, "Set", "app.internal")
+	defer func() {
+		if err != nil {
+			apm.CaptureError(ctx, err).Send()
+		}
```
[Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

AndersonQ commented on x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go:
```diff
@@ -105,25 +121,41 @@ func (e *Controller) Update(c *config.Config) error {
-func (e *Controller) update() error {
+func (e *Controller) update(ctx context.Context) (err error) {
+	span, ctx := apm.StartSpan(ctx, "update", "app.internal")
+	defer func() {
+		if err != nil {
+			apm.CaptureError(ctx, err).Send()
+		}
```
[Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

AndersonQ commented on x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/emitter.go:
```diff
@@ -25,12 +27,20 @@ func New(ctx context.Context, log *logger.Logger, agentInfo *info.AgentInfo, con
+	return func(ctx context.Context, c *config.Config) (err error) {
+		span, ctx := apm.StartSpan(ctx, "update", "app.internal")
+		defer func() {
+			if err != nil {
+				apm.CaptureError(ctx, err).Send()
+			}
```
[Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

AndersonQ commented on x-pack/elastic-agent/pkg/agent/application/pipeline/stream/operator_stream.go:
```diff
@@ -44,8 +48,16 @@ func (b *operatorStream) Specs() map[string]program.Spec {
+func (b *operatorStream) Execute(ctx context.Context, cfg configrequest.Request) (err error) {
+	span, ctx := apm.StartSpan(ctx, "route", "app.internal")
+	defer func() {
+		if err != nil {
+			apm.CaptureError(ctx, err).Send()
+		}
```
[Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

AndersonQ commented on x-pack/elastic-agent/pkg/agent/application/upgrade/step_download.go:
```diff
@@ -20,7 +22,14 @@ import (
+func (u *Upgrader) downloadArtifact(ctx context.Context, version, sourceURI string) (_ string, err error) {
+	span, ctx := apm.StartSpan(ctx, "downloadArtifact", "app.internal")
+	defer func() {
+		if err != nil {
+			apm.CaptureError(ctx, err).Send()
+		}
```
[Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

AndersonQ commented on x-pack/elastic-agent/pkg/agent/cmd/enroll_cmd.go:
```diff
@@ -184,8 +185,16 @@ func newEnrollCmdWithStore(
 	defer c.stopAgent() // ensure its stopped no matter what
+	span, ctx := apm.StartSpan(ctx, "enroll", "app.internal")
+	defer func() {
+		if err != nil {
+			apm.CaptureError(ctx, err).Send()
+		}
```
[Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

AndersonQ commented on x-pack/elastic-agent/pkg/fleetapi/ack_cmd.go:
```diff
@@ -79,24 +81,29 @@ func NewAckCmd(info agentInfo, client client.Sender) *AckCmd {
+	var err error
+	span, ctx := apm.StartSpan(ctx, "execute", "app.internal")
+	defer func() {
+		if err != nil {
+			apm.CaptureError(ctx, err).Send()
+		}
```
[Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

AndersonQ commented on x-pack/elastic-agent/pkg/fleetapi/acker/fleet/fleet_acker.go:
```diff
@@ -48,7 +50,14 @@ func (f *Acker) SetClient(c client.Sender) {
+func (f *Acker) Ack(ctx context.Context, action fleetapi.Action) (err error) {
+	span, ctx := apm.StartSpan(ctx, "ack", "app.internal")
+	defer func() {
+		if err != nil {
+			apm.CaptureError(ctx, err).Send()
+		}
```
[Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

AndersonQ commented on x-pack/elastic-agent/pkg/fleetapi/acker/fleet/fleet_acker.go:
```diff
@@ -69,7 +78,14 @@ func (f *Acker) Ack(ctx context.Context, action fleetapi.Action) error {
+func (f *Acker) AckBatch(ctx context.Context, actions []fleetapi.Action) (err error) {
+	span, ctx := apm.StartSpan(ctx, "ackBatch", "app.internal")
+	defer func() {
+		if err != nil {
+			apm.CaptureError(ctx, err).Send()
+		}
```
[Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

AndersonQ commented on x-pack/elastic-agent/pkg/fleetapi/acker/lazy/lazy_acker.go:
```diff
@@ -36,19 +38,34 @@ func NewAcker(baseAcker batchAcker, log *logger.Logger) *Acker {
+func (f *Acker) Ack(ctx context.Context, action fleetapi.Action) (err error) {
+	span, ctx := apm.StartSpan(ctx, "ack", "app.internal")
+	defer func() {
+		if err != nil {
+			apm.CaptureError(ctx, err).Send()
+		}
```
[Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

AndersonQ commented on x-pack/elastic-agent/pkg/fleetapi/acker/lazy/lazy_acker.go:
```diff
@@ -36,19 +38,34 @@ func NewAcker(baseAcker batchAcker, log *logger.Logger) *Acker {
+func (f *Acker) Commit(ctx context.Context) (err error) {
+	span, ctx := apm.StartSpan(ctx, "commit", "app.internal")
+	defer func() {
+		if err != nil {
+			apm.CaptureError(ctx, err).Send()
+		}
```
[Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

AndersonQ commented on x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go:
```diff
@@ -137,19 +169,21 @@ func (e *Controller) update() error {
 		}
 		err = transpiler.Insert(ast, renderedInputs, "inputs")
 		if err != nil {
-			return errors.New(err, "inserting rendered inputs failed")
```
[Question]
Any reason to remove information from the error?

AndersonQ commented on x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go:
```diff
@@ -164,10 +198,11 @@ func (e *Controller) update() error {
 	}
 
-	return e.router.Route(ast.HashStr(), programsToRun)
+	err = e.router.Route(ctx, ast.HashStr(), programsToRun)
+	return err
```
```suggestion
	return e.router.Route(ctx, ast.HashStr(), programsToRun)
```

AndersonQ commented on x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/emitter.go:
```diff
@@ -25,12 +27,20 @@ func New(ctx context.Context, log *logger.Logger, agentInfo *info.AgentInfo, con
+			}
+			span.End()
+		}()
+		err = ctrl.Update(ctx, c)
+		return err
```
```suggestion
		return ctrl.Update(ctx, c)
```

AndersonQ reviewed (changes requested):
It isn't critical, but I believe the code would profit from the simplifications I suggested.

stuartnelson3 commented on x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go:
```diff
@@ -87,18 +97,19 @@ func (ad *ActionDispatcher) Dispatch(acker store.FleetAcker, actions ...fleetapi
 	}
 
-	return acker.Commit(ad.ctx)
+	err = acker.Commit(ctx)
+	return err
```
cool, didn't know that

stuartnelson3 commented:
@AndersonQ thanks for the suggestions!

axw commented:
@AndersonQ is there anything still needed before we can merge this? Stuart is out this week, but I can try and take a look if there's anything remaining.

AndersonQ reviewed (changes requested):
Hi @axw 

There are a few things I'd like to ask the reasoning behind or to change:
 - closing the default tracer on the init function
 - wrapping the transport where the transport is declared
 - replacing the use of the `:=` notation by expliciting declaring the variable
 - assigning and error to variable just to return the variable.

I've added the comments on the files.

axw commented on x-pack/elastic-agent/pkg/agent/cmd/enroll_cmd.go:
```diff
@@ -219,28 +228,32 @@ func (c *enrollCmd) Execute(ctx context.Context, streams *cli.IOStreams) error {
 	if c.options.DelayEnroll {
 		if c.options.FleetServer.Host != "" {
-			return errors.New("--delay-enroll cannot be used with --fleet-server-es", errors.TypeConfig)
+			err = errors.New("--delay-enroll cannot be used with --fleet-server-es", errors.TypeConfig)
+			return err
```
```suggestion
			return errors.New("--delay-enroll cannot be used with --fleet-server-es", errors.TypeConfig)
```

axw commented on x-pack/elastic-agent/pkg/agent/cmd/enroll_cmd.go:
```diff
@@ -219,28 +228,32 @@ func (c *enrollCmd) Execute(ctx context.Context, streams *cli.IOStreams) error {
 	err = c.enrollWithBackoff(ctx, persistentConfig)
 	if err != nil {
-		return errors.New(err, "fail to enroll")
+		err = errors.New(err, "fail to enroll")
+		return err
```
```suggestion
		return errors.New(err, "fail to enroll")
```

axw commented on x-pack/elastic-agent/pkg/agent/cmd/enroll_cmd.go:
```diff
@@ -219,28 +228,32 @@ func (c *enrollCmd) Execute(ctx context.Context, streams *cli.IOStreams) error {
 		err = install.FixPermissions()
 		if err != nil {
-			return errors.New(err, "failed to fix permissions")
+			err = errors.New(err, "failed to fix permissions")
+			return err
```
```suggestion
			return errors.New(err, "failed to fix permissions")
```

axw commented on x-pack/elastic-agent/pkg/remote/client.go:
```diff
@@ -124,6 +125,8 @@ func NewWithConfig(log *logger.Logger, cfg Config, wrapper wrapperFunc) (*Client
 			}
 		}
 
+		transport = apmhttp.WrapRoundTripper(transport)
+
```
This line can just be removed, the transport is already instrumented by `httpcommon.WithAPMHTTPInstrumentation` above. It just never would have done anything before, as trace context wasn't being propagated.

axw commented on x-pack/elastic-agent/pkg/agent/cmd/run.go:
```diff
@@ -44,6 +47,10 @@ const (
 )
 
+func init() {
+	apm.DefaultTracer.Close()
```
> as apm.DefaultTracer should be closed due to the fact a new tracer is create by NewTracer or NewTracerOptions, better to have them both together. It'll make this relation more explicitly .

The default tracer is not closed because a new one is being created, it's being closed because the default one is not ever expected to be used by Elastic Agent. If it is left open, it will periodically attempt to connect to APM Server. Closing it in package init prevents that as early as possible.

> If in the far future someone else changes the code and decides to use the default tracer, there is a good change they will miss it's being closed here.

We will be releasing a new major version of the agent soon (weeks), and this new version has a more sensible API which does not have side effects due to importing the agent's main package:

```go
package apm

// DefaultTracer returns the default global Tracer, set the first time the
// function is called. It is configured via environment variables.
//
// This will always be initialized to a non-nil value. If any of the
// environment variables are invalid, the corresponding errors will be logged
// to stderr and the default values will be used instead.
func DefaultTracer() *Tracer

// SetDefaultTracer sets the tracer returned by DefaultTracer(). If another
// tracer has already been initialized, it is closed. Any queued events are not
// flushed; it is the responsibility of the caller to call
// DefaultTracer().Flush()
```

I suggest we stick with closing the tracer in package-init for now, and migrate to the new version of the agent ASAP. Then the close call can simply be removed.

axw commented on x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go:
```diff
@@ -87,18 +95,18 @@ func (ad *ActionDispatcher) Dispatch(acker store.FleetAcker, actions ...fleetapi
 	}
 
-	return acker.Commit(ad.ctx)
+	return acker.Commit(ctx)
```
@AndersonQ @ph I just noticed this while looking through the diff. Is this change safe?

AndersonQ replied:
We're on an odd scenario here, there are 2 contexts for this method, this shouldn't happen.
 - `ad.ctx` is for cancelation, it shuts the `ActionDispatcher` down
 - `ctx` is just a carrier for the span.
 
Any proper solution would require a good amount of refactoring. Also passing in just the span isn't ideal.

For simplicity, I suggest to document that, let's comment the method and make clear that for now the only allowed function for the context received as parameter is to carry the span, and that cancelation is done through the `ActionDispatcher` context.

I'm adding my suggestion to fix that

@ph, what do you think?

axw commented:
Thanks @AndersonQ! Agreed on (almost) all points.

AndersonQ commented on x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go:
```diff
@@ -74,7 +76,13 @@ func (ad *ActionDispatcher) key(a fleetapi.Action) string {
 
 // Dispatch dispatches an action using pre-registered set of handlers.
-func (ad *ActionDispatcher) Dispatch(acker store.FleetAcker, actions ...fleetapi.Action) error {
+func (ad *ActionDispatcher) Dispatch(ctx context.Context, acker store.FleetAcker, actions ...fleetapi.Action) (err error) {
+	span, ctx := apm.StartSpan(ctx, "dispatch", "app.internal")
+	defer func() {
+		apm.CaptureError(ctx, err).Send()
+		span.End()
+	}()
+
```
My suggestion to have a context with the `ad.ctx` cancelation and the span. See [this](https://github.com/elastic/beats/pull/30385/files#r813778664) comment for the reasoning behind this.


```suggestion
// ctx is used here ONLY to carry the span, for cancellation use the cancel
// function of the ActionDispatcher.ctx.
func (ad *ActionDispatcher) Dispatch(ctx context.Context, acker store.FleetAcker, actions ...fleetapi.Action) (err error) {
	span, ctx := apm.StartSpan(ctx, "dispatch", "app.internal")
	// Creating a child context that carries both the ad.ctx cancellation and
	// the span from ctx.
	ctx, _ = context.WithCancel(ad.ctx) // ignoring the cancel func as ctx will be cancelled when ever ad.ctx is cancelled.
	ctx = apm.ContextWithSpan(ctx, span)

```

AndersonQ reviewed (changes requested):
@axw [this](https://github.com/elastic/beats/pull/30385/files#r813506319) was a good catch. We need to fix that.

I proposed a solution, let me know what you think.

AndersonQ commented on x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher_test.go:
```diff
@@ -49,8 +49,9 @@ func TestActionDispatcher(t *testing.T) {
 	ack := noopacker.NewAcker()
 
 	t.Run("Success to dispatch multiples events", func(t *testing.T) {
+		ctx := context.Background()
```
Given [my change proposal](https://github.com/elastic/beats/pull/30385/files#r813779788), better to add a test.


This is the type definition, I couldn't add as a change suggestion
```
type mockAcker struct {
	CommitFn func(ctx context.Context) error
}

func (m mockAcker) Ack(ctx context.Context, action fleetapi.Action) error {
	panic("implement me")
}

func (m mockAcker) Commit(ctx context.Context) error {
	return m.CommitFn(ctx)
}
```

```suggestion
	ack := noopacker.NewAcker()

	t.Run("Merges ActionDispatcher ctx cancel and Dispatch cxt value", func(t *testing.T) {
		action1 := &mockAction{}
		def := &mockHandler{}

		span := apmtest.NewRecordingTracer().
			StartTransaction("ignore", "ignore").
			StartSpan("ignore", "ignore", nil)

		ctx1, cancel := context.WithCancel(context.Background())
		ack := mockAcker{CommitFn: func(ctx context.Context) error {
			// ctx1 not cancelled yet
			require.NoError(t, ctx.Err())

			got := apm.SpanFromContext(ctx)
			require.Equal(t, span.TraceContext().Span, got.ParentID())

			cancel() // cancel function from ctx1
			require.Error(t, ctx.Err(), context.Canceled)

			return nil
		}}

		d, err := New(ctx1, nil, def)
		require.NoError(t, err)

		ctx2 := apm.ContextWithSpan(context.Background(), span)
		err = d.Dispatch(ctx2, ack, action1)
		require.NoError(t, err)
	})

	t.Run("Success to dispatch multiples events", func(t *testing.T) {
		ctx := context.Background()
```

stuartnelson3 commented:
@AndersonQ i updated the code as per your suggestions, thank you! the one change I made was wrt discarding the `cancel()` func. golint won't let this happen, so I added a `defer cancel()`. I looked through the code and it looked like this should be fine (it seems each ctx except for `ad.ctx` is scoped to the function call?), but let me know if there's a reason we shouldn't be canceling this sub context!

ph reviewed (approved):
This looks ok on my side, I haven't tested it yet.
Just want to make sure this goes in the next minor correct?

stuartnelson3 commented:
@ph Yes, definitely. And if at all possible, getting it merged before elastic-agent is migrated to its own repo and this PR needs to be re-opened there :)

ph commented:
@AndersonQ Can you do a manual test of this PR so we can unblock @stuartnelson3 ?

AndersonQ commented on x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go:
```diff
@@ -74,7 +76,22 @@ func (ad *ActionDispatcher) key(a fleetapi.Action) string {
+// ctx is used here ONLY to carry the span, for cancelation use the cancel
+// function of the ActionDispatcher.ctx.
+func (ad *ActionDispatcher) Dispatch(ctx context.Context, acker store.FleetAcker, actions ...fleetapi.Action) (err error) {
+	span, ctx := apm.StartSpan(ctx, "dispatch", "app.internal")
```
[Blocker]
@stuartnelson3 the span isn't being ended anymore or capturing the error. Was it intentional?

stuartnelson3 replied:
That was my mistake from cargo-culting your code suggestion above, i'll add it back

AndersonQ commented on x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go:
```diff
@@ -74,7 +76,22 @@ func (ad *ActionDispatcher) key(a fleetapi.Action) string {
+	span, ctx := apm.StartSpan(ctx, "dispatch", "app.internal")
+	// Creating a child context that carries both the ad.ctx cancelation and
+	// the span from ctx.
+	var cancel func()
+	ctx, cancel = context.WithCancel(ad.ctx)
```
If you make changes for the comment above, could you also change here to use `:=`?
Is there a special reason you avoid using `:=`?
```suggestion
	ctx, cancel := context.WithCancel(ad.ctx)
```

stuartnelson3 replied:
no reason, adding it back

stuartnelson3 commented:
@ph @AndersonQ any idea what the failures are? they only started with https://github.com/elastic/beats/pull/30385/commits/61da5fc812f4c144715f82287cd8153b936d66fd but i have no idea why.

from the e2e failure:

```
[2022-03-02T12:04:01.357Z] tar: /tmp/beats-events-20220302.ndjson: Cannot stat: No such file or directory
[2022-03-02T12:04:01.357Z] tar: Exiting with failure status due to previous errors
```

ph commented:
@stuartnelson3 I've never seen this on, not sure what `beats-events*.json`, I will check when this CI run is complete.

stuartnelson3 commented:
/test

mdelapenya commented:
I'd say the error could be related to a network glitch while downloading the elastic-agent from Elastic's artifacts API:

>[2022-03-03T10:59:24.323Z] DEBU[2022-03-03T10:59:21Z] The Elastic artifacts API is available        elapsedTime=141.051415ms retries=1 statusEndpoint="https://artifacts-api.elastic.co/v1/search/8.2.0-fdde08ec/elastic-agent?x-elastic-no-kpi=true"
[2022-03-03T10:59:24.323Z] TRAC[2022-03-03T10:59:21Z] Artifact found                                artifact=elastic-agent artifactName=elastic-agent-8.2.0-SNAPSHOT-linux-arm64.tar.gz elapsedTime=141.343787ms retries=1 version=8.2.0-fdde08ec
[2022-03-03T10:59:24.323Z] TRAC[2022-03-03T10:59:21Z] Downloading file                              path=/tmp/b331ed45-42e7-4c67-97d1-b368749727a7/94807414-5062-4430-9b98-cf6792ba546a url="https://snapshots.elastic.co/8.2.0-fdde08ec/downloads/beats/elastic-agent/elastic-agent-8.2.0-SNAPSHOT-linux-arm64.tar.gz"
[2022-03-03T10:59:24.323Z] TRAC[2022-03-03T10:59:21Z] File downloaded                               elapsedTime=381.276435ms path=/tmp/b331ed45-42e7-4c67-97d1-b368749727a7/94807414-5062-4430-9b98-cf6792ba546a retries=1 url="https://snapshots.elastic.co/8.2.0-fdde08ec/downloads/beats/elastic-agent/elastic-agent-8.2.0-SNAPSHOT-linux-arm64.tar.gz"
[2022-03-03T10:59:28.535Z] ERRO[2022-03-03T10:59:27Z] Could not write file                          error="stream error: stream ID 1; INTERNAL_ERROR" path=/tmp/b331ed45-42e7-4c67-97d1-b368749727a7/94807414-5062-4430-9b98-cf6792ba546a url="https://snapshots.elastic.co/8.2.0-fdde08ec/downloads/beats/elastic-agent/elastic-agent-8.2.0-SNAPSHOT-linux-arm64.tar.gz"
[2022-03-03T10:59:28.535Z] ERRO[2022-03-03T10:59:27Z] Could not download the binary for the Elastic artifact  arch=arm64 artifact=elastic-agent error="stream error: stream ID 1; INTERNAL_ERROR" extension=tar.gz os=linux version=8.2.0-fdde08ec-SNAPSHOT
[2022-03-03T10:59:28.535Z] ERRO[2022-03-03T10:59:27Z] Could not download the binary                 arch=arm64 artifact=elastic-agent error="stream error: stream ID 1; INTERNAL_ERROR" extension=tar.gz os=linux version=8.2.0-fdde08ec-SNAPSHOT

It happens at the very beginning of the execution of that stage, which correlates with the failing scenario (the first one that is executed). After that, further downloads of the binary works.

I'd say that the error is not caused by this PR and could be ignored, as any other execution should work.

stuartnelson3 commented:
@ph @AndersonQ how do you two feel about @mdelapenya 's comment?

this sounds like a transient network error, so I'll re-run to see if it works

stuartnelson3 commented:
/test

simitt commented:
@ph and @AndersonQ please take some time to support getting this merged in **before** the agent is moved to the new repo. This has been ongoing for quite a while and I think the important part is to get the base merged. Any further details could be ironed out on a follow up; I assume you would like to extend the instrumentation over time anyways.

mdelapenya commented:
The current e2e test failure is under our radar: https://github.com/elastic/e2e-testing/pull/2203

simitt commented:
Thanks for the feedback @ph. @stuartnelson3 is on PTO, can you or @AndersonQ please take ownership of merging this in or specifying what exactly is missing?

simitt commented:
I see that the Elastic Agent has moved to it's own repository. @ph can you please own transfering this PR over and get it over the finish line?

stuartnelson3 commented:
my PTO is being shifted slightly so I'll take a look at transferring this PR over today. I'll catch up with @ph if there is anything I need help with when transferring
